src/main/service/listManipulation/crownFirmListService.ts

```diff
@@ -0,0 +1,114 @@
+import moment from 'moment-timezone';
+import {DataManipulationService} from '../dataManipulationService';
+import {DateTimeHelper} from '../../helpers/dateTimeHelper';
+import {CrimeListsService} from './CrimeListsService';
+
+const dataManipulationService = new DataManipulationService();
+const dailyListService = new CrimeListsService();
+const dateTimeHelper = new DateTimeHelper();
+
+export class CrownFirmListService {
+  public timeZone = 'Europe/London';
+
+  /**
+   * Builds list of courthouses, does general formatting to make it appear nice, then is passed to second function which
+   * deals with the splitting it by various dimensions (i.e. courthouse -> day -> courtRoom -> allocated/unallocated.
+   * @param firmList
+   * @param language
+   * @param languageFile
+   */
+  public splitOutFirmListData(firmList: string, language: string, languageFile: string) {
+    const rows = [];
+    const firmListData = JSON.parse(firmList);
+    firmListData['courtLists'].forEach(courtList => {
+      const courtName = courtList['courtHouse']['courtHouseName'];
+      courtList['courtHouse']['courtRoom'].forEach(courtRoom => {
+        courtRoom['session'].forEach(session => {
+          session['sittings'].forEach(sitting => {
+            const sittingDate = moment.utc(sitting['sittingStart']).tz(this.timeZone).format('dddd DD MMMM YYYY');
+            sitting['formattedDuration'] = dateTimeHelper.formatDuration(sitting['durationAsDays'] as number, sitting['durationAsHours'] as number,
+              sitting['durationAsMinutes'] as number, language, languageFile);
+            sitting['hearing'].forEach(hearing => {
+              dailyListService.findLinkedCasesInformation(hearing);
+              hearing['case'].forEach(thisCase => {
+                const formattedName = hearing['defendant'].split(/\s+/);
+                const row = {
+                  courtName: courtName,
+                  sittingDate: sittingDate,
+                  sittingTime: dataManipulationService.publicationTimeInBst(sitting['sittingStart']),
+                  courtRoom: courtRoom['courtRoomName'],
+                  joh: session['formattedJudiciaries'],
+                  durationAsHours: sitting['durationAsHours'],
+                  durationAsMinutes: sitting['durationAsMinutes'],
+                  formattedDuration: sitting['formattedDuration'],
+                  caseNumber: thisCase['caseNumber'],
+                  caseSeparator: thisCase['caseSequenceIndicator'],
+                  linkedCases: thisCase['linkedCases'],
+                  hearingNotes: hearing['listingNotes'],
+                  defendant: formattedName[1].toUpperCase() + ', ' + formattedName[0],
```

Comment:
I am assuming formattedName[1] is always surname and  formattedName[0] is always forename? If these fields can be empty, we are likely to see a redundant comma and the output could be for example:

Furnivall,
, Daniel

...or just a single comma. Might be better to check both surname and forename are present before adding the comma?

Reply:
Fair point! Fixed.